### ISSUE TITLE
NDDataArray.evaluate does not work properly for spatial axes

### ISSUE THREAD

**Issue by registerrier** [gammapy/gammapy#1308]
Looking at the issue of the Background3D (and any other 3D IRF which will come one day or the other).

Assume you have an IRF stored as function of theta and phi. If you call
``NDDataArray.evaluate(theta,phi,E)`` you will get an output with shape ``(n_theta, n_phi, n_E)`` giving the interpolated value for each couple ``(theta_i, phi_j, E_k)``.

This is not the appropriate behavior in this case since you will usually start with an array of ``(x_i, y_j)``.

Also if you work with ``~gammapy.maps``, you will use arrays of coordinates not distributed orthogonally, e.g. in healpix or sparse maps. You want to be able to pass a set of FoV coordinates adapted to your spatial geometry.

Wouldn't it be better/safer to have ``NDDataArray.evaluate`` working directly with input arrays having the same dimensionality as the NDDataArray itself? Or do we need to treat the spatial coordinates differently (as in ``gammapy.maps``)

Opinions? @cdeil @woodmd @leajouvin @joleroi @adonath

**Comment by woodmd** [comment]
The convention used in `gammapy.maps` is that output arrays generated by methods like `evaluate` will have the dimensionality of the input arguments after broadcasting.  Here are some concrete examples for a method `evaluate(x,y,z)`:
```
x : (10,) y: (10,) z : (10,)
output shape : (10,)

x : (1,) y: (10,) z : (10,)
output shape : (10,)

x : (10,) y: (1,) z : (1,)
output shape : (10,)

x : (10,10) y: (1,10) z : (10,)
output shape : (10,10)
```
Note that in `gammapy.maps` the arguments are actually wrapped into a single tuple/coord object but the approach is the same nonetheless.  I believe this is the convention used in many other packages based on numpy.  What's nice about this approach is that it gives you a lot of flexibility in how you pass values.  For instance if your arguments are orthogonal then you can pass in a sparse grid (e.g. as generated with `np.ogrid`).  For non-orthogonal arguments (e.g. as in spatial coordinates of a healpix map) you would pass a dense grid.

**Comment by cdeil** [comment]
Why is this working https://github.com/gammapy/gammapy/blob/2fac9a63bc26771e615d12065901db67b97ccfb4/gammapy/cube/tests/test_exposure.py#L56
I think maybe because there was a 2D array `detx` and a scalar `dety` and a 1D array `energy` and then Numpy broadcasting does it's magic? And when you pass a 2D array for `dety` also it blows up and tries to evaluate on a 5D grid?

I think for now you can avoid it by passing 3D arrays for all parameters and then internally no broadcasting happens?

It's not clear to me overall if we should introduce a new class for `FOVCoordinate`, or use `SkyCoordinate` with an "offset" frame for that, or use `gammapy.map.MapCoords`. Clearly the code to do transforms between sky and FOV coordinates should be in a single place, and probably not in the IRF class? So maybe a separate class makes sense, and then FOV coordinates are another special king of coordinate that maps and IRFs know a little bit about?

If someone has a proposal, please write it here.
Otherwise, let's discuss tomorrow in the call?

**Comment by cdeil** [comment]
No matter how the book-keeping is done, the coordinate computations would start with 2D grids of lon / lat on the sky via some property or method call on map geom, then go through transforms to get a 2D grid of coords that IRF evaluate understands (in this case for Background3D detx / dety).

**Comment by cdeil** [comment]
Just so that we have a common test case ...

I tried this and indeed it leads to 100% CPU and 40 GB of RAM use and I had to kill the process:
```python
import numpy as np
import astropy.units as u
from gammapy.irf import Background3D
filename = '$GAMMAPY_EXTRA/datasets/cta-1dc/caldb/data/cta//1dc/bcf/South_z20_50h/irf_file.fits'
bkg_3d = Background3D.read(filename, hdu='BACKGROUND')
print(bkg_3d)
detx = np.zeros((200, 100)) * u.deg
dety = np.zeros((200, 100)) * u.deg
energy = np.arange(1, 10) * u.TeV
bkg_3d.data.evaluate(detx=detx, dety=dety, energy=energy)
```
Makes sense, the class currently has no knowledge that detx and dety are connected, i.e. should get special treatment in broadcasting (in this case I want a (200, 100, 9) output array).

**Comment by JouvinLea** [comment]
@woodmd 
In your example, would it work if we have:
```x : (10,10) y: (10,10) z : (10,)```
to have the ```output shape : (10,10,10)```
For the moment the evaluate method of the NDdataArray will return (10,10,10,10,10)....

@cdeil 
There was no issue for gammapy/gammapy/cube/tests/test_exposure.py since the IRF contains only one axis of spatial coordinates. You will alway evaluate only on offset and energy that are orthogonal coordinates so there was no issue as for all the IRFs we were using until now. The issue appeared as @registerrier said for the `Background 3D` class presenting 2 axes for spatial coordinates.

@cdeil : I think a `FOVCoordinate` class will be useful to be abble to play as the SKyCoord. However I don't see the link with this issue ?

@cdeil ```I think for now you can avoid it by passing 3D arrays for all parameters and then internally no broadcasting happens?```
I don't think it works. I tried passing for the ```BAckground3D``` 3 arrays of shape (3,5,6) for detx, dety and energy, and it returns a shape (4, 5, 6, 4, 5, 6) after `evaluate()`

Would it be possible in the class NDataArray to not broadcast for the axes that are spatial coordinates?

**Comment by cdeil** [comment]
> I don't think it works. I 

You are right. Just to give an example to reproduce:
```
>>> import numpy as np
>>> import astropy.units as u
>>> from gammapy.irf import Background3D
>>> filename = '$GAMMAPY_EXTRA/datasets/cta-1dc/caldb/data/cta//1dc/bcf/South_z20_50h/irf_file.fits'
>>> bkg_3d = Background3D.read(filename, hdu='BACKGROUND')
>>> print(bkg_3d)
Background3D
NDDataArray summary info
energy         : size =    21, min =  0.016 TeV, max = 158.489 TeV
detx           : size =    36, min = -5.833 deg, max =  5.833 deg
dety           : size =    36, min = -5.833 deg, max =  5.833 deg
Data           : size = 27216, min =  0.000 1 / (MeV s sr), max =  0.421 1 / (MeV s sr)

>>> detx = np.zeros((2, 3, 4)) * u.deg
>>> dety = np.zeros((2, 3, 4)) * u.deg
>>> energy = np.ones((2, 3, 4)) * u.TeV
>>> val = bkg_3d.data.evaluate(detx=detx, dety=dety, energy=energy)
>>> val.shape
(2, 3, 4, 2, 3, 4, 2, 3, 4)
```

> @cdeil : I think a FOVCoordinate class will be useful to be abble to play as the SKyCoord. However I don't see the link with this issue ?
> Would it be possible in the class NDataArray to not broadcast for the axes that are spatial coordinates?

This is the link. If we want special behaviour for these two coupled axes, we have to introduce something that couples the axes.

In addition to what we currently have in `gammapy.utils.nddata` for ND array and axes, and what we have in gammapy.maps for maps and axes and WCS, we should probably look at these two examples how to handle ND including a system to have "coupled" axed like lon / lat:

* https://github.com/astrofrog/astropy-APEs/blob/29ecc33da866dc8d58175c38550948b48ee98cef/APE14.rst
* http://xarray.pydata.org/en/stable/examples/multidimensional-coords.html

I'm adding this to the agenda for the call tonight now:
https://github.com/gammapy/gammapy-meetings/tree/master/2018-02-15
For me, actually this question of how coordinates are handled is the biggest / most pressing remaining point to discuss.

**Comment by cdeil** [comment]
For gamma.maps, the coordinate interface was improved in #1318 
For IRFs, we'll have to see.

Just a note: I think looking at http://xarray.pydata.org/en/stable/indexing.html and either using xarray for IRFs directly or re-implementing a small subset of what they have done, that we need for our application, would be a good idea.

**Comment by cdeil** [comment]
Fixed by @leajouvin  in #1472 . Closing.
